sdk/typescript/src/rpc/websocket-client.ts

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { isSubscriptionEvent } from '../types/index.guard';
+import {
+  SuiEventFilter,
+  SuiEventEnvelope,
+  SubscriptionId,
+} from '../types';
+import { Client as WsRpcClient} from 'rpc-websockets';
+
+
+export const getWebsocketUrl = (httpUrl: string, port?: number): string => {
+  const url = new URL(httpUrl);
+  url.protocol = url.protocol.replace('http', 'ws');
+  url.port = (port ?? 9001).toString();
+  return url.toString();
+};
+
+enum ConnectionState {
+  NotConnected,
+  Connecting,
+  Connected
+}
+
+type JsonRpcMethodMessage<T> = {
+  jsonrpc: '2.0',
+  method: string,
+  params: T
+}
+
+type FilterSubHandler = {
+  id: SubscriptionId,
+  onMessage: (event: SuiEventEnvelope) => void,
+  filter: SuiEventFilter
+};
+
+type SubscriptionData = {
+  filter: SuiEventFilter,
+  onMessage: (event: SuiEventEnvelope) => void
+}
+
+type MinimumSubscriptionMessage = {
+  subscription: SubscriptionId,
+  result: object
+}
+
+const isMinimumSubscriptionMessage = (msg: any): msg is MinimumSubscriptionMessage =>
+  msg
+  && ('subscription' in msg && typeof msg['subscription'] === 'number')
+  && ('result' in msg && typeof msg['result'] === 'object')
+
+export type WebsocketClientOptions = {
```

Comment:
Can we add more documentation about what each of the parameters means here? What's the expected behavior if the connection get lost and what should the caller do?